File: packages/pigeon/platform_tests/windows_unit_tests/windows/test/multiple_arity_test.cpp
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <gtest/gtest.h>

#include "test/multiple_arity.g.h"
#include "test/utils/fake_host_messenger.h"

namespace multiple_arity_pigeontest {

namespace {
using flutter::EncodableList;
using flutter::EncodableMap;
using flutter::EncodableValue;
using testing::FakeHostMessenger;

class TestHostApi : public MultipleArityHostApi {
 public:
  TestHostApi() {}
  virtual ~TestHostApi() {}

 protected:
  ErrorOr<int64_t> Subtract(int64_t x, int64_t y) override { return x - y; }
};

const EncodableValue& GetResult(const EncodableValue& pigeon_response) {
  return std::get<EncodableMap>(pigeon_response).at(EncodableValue("result"));
}
}  // namespace

TEST(MultipleArity, HostSimple) {
  FakeHostMessenger messenger(&MultipleArityHostApi::GetCodec());
  TestHostApi api;
  MultipleArityHostApi::SetUp(&messenger, &api);

  int64_t result = 0;
  messenger.SendHostMessage("dev.flutter.pigeon.MultipleArityHostApi.subtract",
                            EncodableValue(EncodableList({
                                EncodableValue(30),
                                EncodableValue(10),
                            })),
                            [&result](const EncodableValue& reply) {
                              result = GetResult(reply).LongValue();
                            });

  EXPECT_EQ(result, 20);
}

// TODO(stuartmorgan): Add a FlutterApi version of the test.

}  // namespace multiple_arity_pigeontest

File: packages/pigeon/platform_tests/shared_test_plugin_code/test/multiple_arity_test.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/annotations.dart';
import 'package:mockito/mockito.dart';
import 'package:shared_test_plugin_code/src/generated/multiple_arity.gen.dart';

import 'multiple_arity_test.mocks.dart';

@GenerateMocks(<Type>[BinaryMessenger])
void main() {
  test('multiple arity', () async {
    final BinaryMessenger mockMessenger = MockBinaryMessenger();
    when(
      mockMessenger.send(
        'dev.flutter.pigeon.pigeon_integration_tests.MultipleArityHostApi.subtract',
        any,
      ),
    ).thenAnswer((Invocation realInvocation) async {
      final Object input =
          MultipleArityHostApi.pigeonChannelCodec.decodeMessage(
            realInvocation.positionalArguments[1] as ByteData?,
          )!;
      final List<Object?> args = input as List<Object?>;
      final int x = (args[0] as int?)!;
      final int y = (args[1] as int?)!;
      return MultipleArityHostApi.pigeonChannelCodec.encodeMessage(<Object>[
        x - y,
      ]);
    });

    final MultipleArityHostApi api = MultipleArityHostApi(
      binaryMessenger: mockMessenger,
    );
    final int result = await api.subtract(30, 10);
    expect(result, 20);
  });
}
